radixdlt/src/main/java/com/radixdlt/consensus/BFTEventPreprocessor.java

```diff
@@ -0,0 +1,221 @@
+		if (!Objects.equals(proposerElection.getProposer(view), myKey)) {
+			log.warn("{}: VOTE: Ignoring confused vote {} for {}",
+				getShortName(), vote.hashCode(), vote.getVoteData().getProposed().getView());
+			return;
+		}
+
+		forwardTo.processVote(vote);
+	}
+
+	private boolean processNewViewInternal(NewView newView) {
+		log.trace("{}: NEW_VIEW: PreProcessing {}", getShortName(), newView);
+
+		// only do something if we're actually the leader for the view
+		final View view = newView.getView();
+		if (!Objects.equals(proposerElection.getProposer(view), myKey)) {
+			log.warn("{}: NEW_VIEW: Got confused new-view {} for view {}", getShortName(), newView.hashCode(), newView.getView());
+			return true;
+		}
+
+		final View currentView = pacemakerState.getCurrentView();
+		if (newView.getView().compareTo(currentView) < 0) {
+			log.trace("{}: NEW_VIEW: Ignoring {} Current is: {}", getShortName(), newView.getView(), currentView);
+			return true;
+		}
+
+		if (this.vertexStore.syncToQC(newView.getQC())) {
+			forwardTo.processNewView(newView);
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public void processNewView(NewView newView) {
+		log.trace("{}: NEW_VIEW: Queueing {}", this.getShortName(), newView);
+		if (queues.checkOrAdd(newView)) {
+			if (!processNewViewInternal(newView)) {
+				log.info("{}: NEW_VIEW: Queuing {} Waiting for Sync", getShortName(), newView);
+				queues.add(newView);
+			}
+		}
+	}
+
+	private boolean processProposalInternal(Proposal proposal) {
+		log.trace("{}: PROPOSAL: PreProcessing {}", this.getShortName(), proposal);
+
+		final Vertex proposedVertex = proposal.getVertex();
+		final View proposedVertexView = proposedVertex.getView();
+		final View currentView = this.pacemakerState.getCurrentView();
+		if (proposedVertexView.compareTo(currentView) < 0) {
+			log.trace("{}: PROPOSAL: Ignoring view {} Current is: {}", this.getShortName(), proposedVertexView, currentView);
+			return true;
+		}
+
+		if (this.vertexStore.syncToQC(proposedVertex.getQC())) {
+			forwardTo.processProposal(proposal);
+			if (vertexStore.getVertex(proposedVertex.getId()) != null) {
+				processLocalSync(proposal.getVertex().getId());
+			}
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	@Override
+	public void processProposal(Proposal proposal) {
+		log.trace("{}: PROPOSAL: Queueing {}", this.getShortName(), proposal);
+		if (queues.checkOrAdd(proposal)) {
+			if (!processProposalInternal(proposal)) {
+				log.info("{}: PROPOSAL: Queuing {} Waiting for Sync", getShortName(), proposal);
+				queues.add(proposal);
+			}
+		}
+	}
+
+	@Override
+	public void processLocalTimeout(View view) {
+		final View curView = this.pacemakerState.getCurrentView();
+		forwardTo.processLocalTimeout(view);
+		final View nextView = this.pacemakerState.getCurrentView();
+		if (!curView.equals(nextView)) {
+			// Could probably forward some of these but don't worry for now
```

Comment:
Why would we want to forward some timeouts?

Reply:
bad wording